File: src/aerospike_py/_bug_report.py
"""Internal bug report helper for unexpected errors in aerospike-py."""

from __future__ import annotations

import asyncio
import functools
import logging
import platform
import sys
import traceback

logger = logging.getLogger("aerospike_py")

_REPO = "KimSoungRyoul/aerospike-py"


def _shell_escape(s: str) -> str:
    """Escape single quotes for shell single-quoted strings."""
    return s.replace("'", "'\\''")


def log_unexpected_error(context: str, exc: BaseException) -> None:
    """Log an unexpected error with a gh issue create command.

    Only call this for errors that are NOT expected Aerospike errors
    (i.e., not subclasses of AerospikeError).
    """
    from aerospike_py import __version__

    exc_type = type(exc).__name__
    exc_msg = str(exc)
    tb_str = "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)).strip()

    title = f"Unexpected error: {exc_type}: {exc_msg[:80]}"
    body = (
        f"aerospike-py version: {__version__}\n"
        f"Python: {sys.version}\n"
        f"Platform: {platform.platform()}\n"
        f"Context: {context}\n"
        f"Error: {exc_type}: {exc_msg}\n\n"
        f"Traceback:\n```\n{tb_str}\n```"
    )

    logger.error(
        "Unexpected internal error in aerospike-py (%s): %s: %s\n"
        "\n"
        "This error may be a bug in aerospike-py. Please report it by running:\n"
        "gh issue create --repo %s --title '%s' --body '%s'",
        context,
        exc_type,
        exc_msg,
        _REPO,
        _shell_escape(title),
        _shell_escape(body),
    )


def _maybe_log(method_name: str, exc: Exception) -> None:
    """Log if exc is NOT an expected AerospikeError."""
    from aerospike_py._aerospike import AerospikeError

    if not isinstance(exc, AerospikeError):
        log_unexpected_error(method_name, exc)


def catch_unexpected(method_name: str):
    """Decorator that catches unexpected errors and logs a bug report.

    Expected AerospikeError subclasses pass through unmodified.
    All other exceptions are logged with the bug report message,
    then re-raised as-is (so the caller still sees the original error).
    """

    def decorator(func):
        if asyncio.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args, **kwargs):
                try:
                    return await func(*args, **kwargs)
                except Exception as exc:
                    _maybe_log(method_name, exc)
                    raise

            return async_wrapper
        else:

            @functools.wraps(func)
            def wrapper(*args, **kwargs):
                try:
                    return func(*args, **kwargs)
                except Exception as exc:
                    _maybe_log(method_name, exc)
                    raise

            return wrapper

    return decorator

File: rust/src/bug_report.rs
//! Bug report logging for unexpected internal errors.
//!
//! When aerospike-py encounters an error that is likely a library bug
//! (not an expected Aerospike server error), these helpers log a message
//! suggesting the user file a GitHub issue.

use log::error;

const REPO: &str = "KimSoungRyoul/aerospike-py";

/// Escape single quotes for shell single-quoted strings: `'` → `'\''`.
fn shell_escape(s: &str) -> String {
    s.replace('\'', "'\\''")
}

/// Truncate and sanitize a string for use in a GitHub issue title.
fn sanitize_for_title(s: &str, max_len: usize) -> String {
    let cleaned: String = s.chars().map(|c| if c == '\n' { ' ' } else { c }).collect();
    if cleaned.len() <= max_len {
        cleaned
    } else {
        format!("{}...", &cleaned[..max_len - 3])
    }
}

/// Log an unexpected internal error with a `gh issue create` command.
///
/// Only call this for errors that are NOT expected Aerospike errors.
pub fn log_unexpected_error(context: &str, error_detail: &str) {
    let version = env!("CARGO_PKG_VERSION");
    let title = format!("Unexpected error: {}", sanitize_for_title(error_detail, 80));
    let body =
        format!("aerospike-py version: {version}\nContext: {context}\nError: {error_detail}");
    error!(
        "Unexpected internal error in aerospike-py ({context}): {error_detail}\n\
         \n\
         This error may be a bug in aerospike-py. Please report it by running:\n\
         gh issue create --repo {REPO} \
         --title '{title}' \
         --body '{body}'",
        title = shell_escape(&title),
        body = shell_escape(&body),
    );
}

/// Replace `unreachable!()` with a version that logs a bug report
/// and returns a `PyErr` instead of panicking.
macro_rules! internal_bug {
    ($context:expr, $($arg:tt)*) => {{
        let detail = format!($($arg)*);
        $crate::bug_report::log_unexpected_error($context, &detail);
        Err(pyo3::exceptions::PyRuntimeError::new_err(
            format!("aerospike-py internal error ({}): {}", $context, detail)
        ))
    }};
}

pub(crate) use internal_bug;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_short_string() {
        assert_eq!(sanitize_for_title("hello", 80), "hello");
    }

    #[test]
    fn test_sanitize_long_string() {
        let long = "a".repeat(100);
        let result = sanitize_for_title(&long, 80);
        assert_eq!(result.len(), 80);
        assert!(result.ends_with("..."));
    }

    #[test]
    fn test_sanitize_newlines() {
        assert_eq!(sanitize_for_title("line1\nline2", 80), "line1 line2");
    }

    #[test]
    fn test_sanitize_exact_length() {
        let exact = "a".repeat(80);
        assert_eq!(sanitize_for_title(&exact, 80), exact);
    }

    #[test]
    fn test_shell_escape_no_quotes() {
        assert_eq!(shell_escape("hello world"), "hello world");
    }

    #[test]
    fn test_shell_escape_with_quotes() {
        assert_eq!(shell_escape("it's a bug"), "it'\\''s a bug");
    }
}
